### PR TITLE
feat: add hashtags to event cards

### LINKED ISSUE #1592: Add hashtags to event cards
**Describe the feature you'd like**
Add hashtags as chips to event cards similar to the website 

**Would you like to work on the issue?**
- [x] Yes

### PR DESCRIPTION
**Pull request by aggarwalpulkit596**
Fixes #1592

Changes: Added Event Types, SubTopics to the database to show hashtags in event cards

Screenshots for the change:
screenshot-1555442322949

### CODE REVIEW

File: app/src/main/java/org/fossasia/openevent/general/event/EventService.kt [hunk 1]
```diff
@@ -76,6 +94,8 @@ class EventService(
         return eventApi.searchEvents("name", locationName).flatMap { apiList ->
             val eventIds = apiList.map { it.id }.toList()
             eventTopicsDao.insertEventTopics(getEventTopicList(apiList))
+            eventTypesDao.insertEventTypes(getEventTypeList(apiList))
+            eventSubTopicDao.insertEventSubTopics(getEventSubTopicList(apiList))
             eventDao.getFavoriteEventWithinIds(eventIds).flatMap { favIds ->
```
[the remark this answers is not in the document]
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/EventService.kt]: Include those in the event call

[the remark this answers is not in the document]
Comment (aggarwalpulkit596) [reply app/src/main/java/org/fossasia/openevent/general/event/EventService.kt]: Sorry i didn't get you how should i include event call

[the remark this answers is not in the document]
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/EventService.kt]: See JSON:API spec, you can include related objects in response

**Comment by aggarwalpulkit596** [comment]
I tried using resolve = true but it isn't Saving the object

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal @nikit19 updated please review

**Comment by iamareebjamal** [comment]
Use the outline style in chip

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal i couldn't find any chip with name outline can you choose from one of them

style="@style/Widget.MaterialComponents.Chip.Entry"
style="@style/Widget.MaterialComponents.Chip.Filter"
style="@style/Widget.MaterialComponents.Chip.Choice"
style="@style/Widget.MaterialComponents.Chip.Action"

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal updated

**Comment by iamareebjamal** [comment]
Peer Review

**Comment by aggarwalpulkit596** [comment]
@liveHarshit please review

**Comment by liveHarshit** [comment]
UI is not looking good. You can decrease the size to make UI effective.

**Comment by aggarwalpulkit596** [comment]
okay anything related to code @liveHarshit ?

File: app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt [hunk 2]
```diff
@@ -27,6 +27,7 @@ import kotlinx.android.synthetic.main.content_event.similarEventsContainer
 import kotlinx.android.synthetic.main.content_event.view.eventDateDetailsFirst
 import kotlinx.android.synthetic.main.content_event.view.eventDateDetailsSecond
 import kotlinx.android.synthetic.main.content_event.view.eventDescription
+import kotlinx.android.synthetic.main.content_event.view.eventImage
```
Comment (liveHarshit) [inline app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt]: Unrelated changes

File: app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt [hunk 3]
```diff
@@ -35,18 +36,17 @@ import kotlinx.android.synthetic.main.content_event.view.eventOrganiserName
 import kotlinx.android.synthetic.main.content_event.view.eventTimingLinearLayout
 import kotlinx.android.synthetic.main.content_event.view.imageMap
 import kotlinx.android.synthetic.main.content_event.view.locationUnderMap
-import kotlinx.android.synthetic.main.content_event.view.eventImage
-import kotlinx.android.synthetic.main.content_event.view.organizerLogoIcon
 import kotlinx.android.synthetic.main.content_event.view.nestedContentEventScroll
+import kotlinx.android.synthetic.main.content_event.view.organizerContainer
+import kotlinx.android.synthetic.main.content_event.view.organizerLogoIcon
```
Comment (liveHarshit) [inline app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt]: same

File: app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt [hunk 4]
```diff
@@ -35,18 +36,17 @@ import kotlinx.android.synthetic.main.content_event.view.eventOrganiserName
...
 import kotlinx.android.synthetic.main.content_event.view.organizerName
 import kotlinx.android.synthetic.main.content_event.view.refundPolicy
 import kotlinx.android.synthetic.main.content_event.view.seeMore
 import kotlinx.android.synthetic.main.content_event.view.seeMoreOrganizer
-import kotlinx.android.synthetic.main.content_event.view.organizerContainer
+import kotlinx.android.synthetic.main.content_fetching_event_error.view.retry
 import kotlinx.android.synthetic.main.fragment_event.view.buttonTickets
-import kotlinx.android.synthetic.main.fragment_event.view.eventErrorCard
 import kotlinx.android.synthetic.main.fragment_event.view.container
-import kotlinx.android.synthetic.main.content_fetching_event_error.view.retry
```
Comment (liveHarshit) [inline app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt]: same

File: app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt [hunk 5]
```diff
@@ -57,14 +57,13 @@ import org.fossasia.openevent.general.ticket.TicketsFragmentArgs
...
 import org.fossasia.openevent.general.utils.Utils.getAnimSlide
 import org.fossasia.openevent.general.utils.Utils.requireDrawable
+import org.fossasia.openevent.general.utils.Utils.setToolbar
 import org.fossasia.openevent.general.utils.extensions.nonNull
 import org.fossasia.openevent.general.utils.nullToEmpty
 import org.fossasia.openevent.general.utils.stripHtml
 import org.koin.androidx.viewmodel.ext.android.viewModel
 import timber.log.Timber
 import java.util.Currency
-import org.fossasia.openevent.general.utils.Utils.setToolbar
```
Comment (liveHarshit) [inline app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt]: same

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt]: And many more in this file

**Comment by aggarwalpulkit596** [comment]
> UI is not looking good. You can decrease the size to make UI effective.

@liveHarshit which one you're referring to outline or purple colored ones?

**Comment by liveHarshit** [comment]
> > UI is not looking good. You can decrease the size to make UI effective.
> 
> @liveHarshit which one you're referring to outline or purple colored ones?

The overall size of the chips.

**Comment by aggarwalpulkit596** [comment]
@liveHarshit updated

**Comment by aggarwalpulkit596** [comment]
@liveHarshit any suggestion which library to use instead of com.google.code.gson:gson:2.8.5to work with both of them

**Comment by aggarwalpulkit596** [comment]
@liveHarshit and also build is passing on local and it was also passing before i don't what happened

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal any suggestions for this?

**Comment by iamareebjamal** [comment]
The error is unrelated to GSON, and we are not even using GSON

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal yes we are using to store topic,type and sub topic in event model

File: app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicIdConverter.kt [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package org.fossasia.openevent.general.event.subtopic
+
+import androidx.room.TypeConverter
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+
+class EventSubTopicIdConverter {
```
Comment (aggarwalpulkit596) [inline app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicIdConverter.kt]: @iamareebjamal here i save an embedded object into event model

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicIdConverter.kt]: You have to add a type converter. Room does not know how to save that into DB

**Comment by iamareebjamal** [comment]
If we are using it, then please remove it as we already have GSON

File: app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicIdConverter.kt (hunk 6 above)
Comment (aggarwalpulkit596) [reply app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicIdConverter.kt]: this is an type converter

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal here gson is used to
A Java serialization/deserialization library to convert Java Objects into JSON and back

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal i'm extremely sorry it was my mistake i forgot to push the build.gradle file in which i added a library

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal @liveHarshit updated it should work now

File: app/build.gradle [hunk 7]
```diff
@@ -118,6 +118,7 @@ dependencies {
     testImplementation "androidx.room:room-testing:${roomVersion}"
     implementation 'androidx.preference:preference:1.1.0-alpha01'
     implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+    implementation "com.google.code.gson:gson:2.8.5"
```
Comment (iamareebjamal) [inline app/build.gradle]: Remove

Comment (aggarwalpulkit596) [reply app/build.gradle]: type converter can't work without them

Comment (iamareebjamal) [reply app/build.gradle]: What is GSON?

Comment (iamareebjamal) [reply app/build.gradle]: And why do you need it when the project is using Jackson

Comment (aggarwalpulkit596) [reply app/build.gradle]: GSON
A Java serialization/deserialization library to convert Java Objects into JSON and back
Refrences
https://medium.com/@amit.bhandari/storing-java-objects-other-than-primitive-types-in-room-database-11e45f4f6d22

Comment (iamareebjamal) [reply app/build.gradle]: > Why do you need it when the project is using Jackson?

Comment (aggarwalpulkit596) [reply app/build.gradle]: i need it to convert event topic,subtopic and type model into event model by converting into gson and then converting it back to respective models

File: app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package org.fossasia.openevent.general.event.subtopic
+
+import androidx.room.TypeConverter
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+
+class EventSubTopicConverter {
```
Comment (aggarwalpulkit596) [inline app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt]: @iamareebjamal see this class how i am converting subtopic model

File: app/build.gradle (hunk 7 above)
Comment (iamareebjamal) [reply app/build.gradle]: Why GSON?

Comment (iamareebjamal) [reply app/build.gradle]: Why do you need it when the project is using Jackson?

Comment (aggarwalpulkit596) [reply app/build.gradle]: @iamareebjamal please read this https://medium.com/@amit.bhandari/storing-java-objects-other-than-primitive-types-in-room-database-11e45f4f6d22 you'll get it

Comment (aggarwalpulkit596) [reply app/build.gradle]: i needed event topic models inside event model to save non primitive data type we type converter and gson helps these type converter

Comment (iamareebjamal) [reply app/build.gradle]: But, why do you need it when the project is using Jackson?

Comment (iamareebjamal) [reply app/build.gradle]: Do you know what Jackson is? I'm not talking about the King of Pop here

Comment (aggarwalpulkit596) [reply app/build.gradle]: yes i know Jackson but with Jackson i can't implement the converter part

Comment (aggarwalpulkit596) [reply app/build.gradle]: class EventTypeConverter {

     @TypeConverter
    fun toEventSubTopic(json: String): EventType? {
        val type = object : TypeToken<EventType>() {}.type
        return Gson().fromJson(json, type)
    }

     @TypeConverter
    fun toJson(eventSubTopic: EventType?) = Gson().toJson(eventSubTopic) }
here the implementation can't be achieved using jackson

Comment (iamareebjamal) [reply app/build.gradle]: Why can't it be done using Jackson?

Comment (aggarwalpulkit596) [reply app/build.gradle]: i couldn't find any methods for such thing
https://stackoverflow.com/questions/49064618/typetoken-getparameterized-in-jackson-json

Comment (iamareebjamal) [reply app/build.gradle]: I don't think I have a job of pointing to the documentation of a JSON serializing library

http://lmgtfy.com/?q=convert+object+jackson

Comment (aggarwalpulkit596) [reply app/build.gradle]: @iamareebjamal i'm sorry and intiallly thought you weren't getting the point of gson/jackson

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal updated.Thank you

File: app/src/main/java/org/fossasia/openevent/general/OpenEventDatabase.kt [hunk 9]
```diff
@@ -42,5 +50,9 @@ abstract class OpenEventDatabase : RoomDatabase() {
 
     abstract fun eventTopicsDao(): EventTopicsDao
 
+    abstract fun eventTypesDao(): EventTypesDao
+
+    abstract fun eventSubTopicDao(): EventSubTopicDao
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/OpenEventDatabase.kt]: Since these are now stored with the event, this is unneeded

Comment (aggarwalpulkit596) [reply app/src/main/java/org/fossasia/openevent/general/OpenEventDatabase.kt]: but we can use dao in future for chips or filtering i suggest we should keep them

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/OpenEventDatabase.kt]: We can't use them unless there is a table for them

File: app/src/main/java/org/fossasia/openevent/general/event/EventService.kt [hunk 10]
```diff
@@ -76,6 +94,8 @@ class EventService(
         return eventApi.searchEvents("name", locationName).flatMap { apiList ->
             val eventIds = apiList.map { it.id }.toList()
             eventTopicsDao.insertEventTopics(getEventTopicList(apiList))
+            eventTypesDao.insertEventTypes(getEventTypeList(apiList))
```
Comment (aggarwalpulkit596) [inline app/src/main/java/org/fossasia/openevent/general/event/EventService.kt]: @iamareebjamal i'm saving individually as well so there is a table for them as well

File: app/src/main/java/org/fossasia/openevent/general/OpenEventDatabase.kt (hunk 9 above)
Comment (aggarwalpulkit596) [reply app/src/main/java/org/fossasia/openevent/general/OpenEventDatabase.kt]: i'm saving them as well in a seperate table i could remove that part though but we may need them in future

File: app/src/main/java/org/fossasia/openevent/general/event/EventService.kt (hunk 10 above)
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/EventService.kt]: Why are you converting in JSON then? Just save them as JSON. No need to add tables

Comment (aggarwalpulkit596) [reply app/src/main/java/org/fossasia/openevent/general/event/EventService.kt]: okay i'll remove them

Comment (aggarwalpulkit596) [reply app/src/main/java/org/fossasia/openevent/general/event/EventService.kt]: updated

**Comment by iamareebjamal** [comment]
Since new columns have been added to the table, there must be changes in schema JSON

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal updated JSON schema

**Comment by aggarwalpulkit596** [comment]
Resolved Conflicts

File: app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package org.fossasia.openevent.general.event.subtopic
+
+import androidx.room.TypeConverter
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+
+class EventSubTopicConverter {
+    @TypeConverter
+    fun toEventSubTopic(json: String): EventSubTopic? {
+        return jacksonObjectMapper().readerFor(EventSubTopic::class.java).readValue<EventSubTopic>(json)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt]: There's already an object mapper in dependency graph, use that

File: app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt [hunk 12]
```diff
@@ -0,0 +1,17 @@
+package org.fossasia.openevent.general.event.subtopic
+
+import androidx.room.TypeConverter
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+
+class EventSubTopicConverter {
+    @TypeConverter
+    fun toEventSubTopic(json: String): EventSubTopic? {
+        val objectMapper = ObjectMapper()
```
Comment (aggarwalpulkit596) [inline app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt]: @iamareebjamal updated

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt]: No, this is still instantiating ObjectMapper yourself

Comment (aggarwalpulkit596) [reply app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt]: @iamareebjamal 
     val mapType = object : TypeReference<EventSubTopic>() {}
        return ObjectMapper().readValue(json, mapType)
like this ?
instead of
        val objectMapper = ObjectMapper()
        val mapType = object : TypeReference<EventSubTopic>() {}
        return ObjectMapper().readValue(json, mapType)

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt]: No, you need to use the ObjectMapper from DI. Who said anything about TypeReference or that your implementation was wrong

Comment (aggarwalpulkit596) [reply app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt]: @iamareebjamal how i can pass that in type converter because type converter can only have empty constructors

Comment (aggarwalpulkit596) [reply app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt]: and also ListAttendeeIdConverter uses the same pattern because we can't pass anything to type converter

Comment (aggarwalpulkit596) [reply app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt]: @iamareebjamal anything else?

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal please review

File: app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt (hunk 12 above)
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt]: You can manually inject the dependency. See field injection

Comment (aggarwalpulkit596) [reply app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt]: but for field injection, we need a context or something type converter don't even have that
i tired using  StandAloneContext.getKoin().koinContext and         val objectMapper: ObjectMapper by inject()

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/subtopic/EventSubTopicConverter.kt]: OK. Create an issue to convert all object mappers from the one in DI

File: app/src/main/java/org/fossasia/openevent/general/event/types/EventTypeConverter.kt [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+import androidx.room.TypeConverter
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+
+class EventTypeConverter {
+    @TypeConverter
+    fun toEventType(json: String): EventType? {
+        val mapType = object : TypeReference<EventType>() {}
+        return ObjectMapper().readValue(json, mapType) }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/event/types/EventTypeConverter.kt]: Revert to how it was before

**Comment by aggarwalpulkit596** [comment]
@iamareebjamal updated

File: app/src/main/java/org/fossasia/openevent/general/event/EventViewHolder.kt [hunk 14]
```diff
@@ -57,6 +61,19 @@ class EventViewHolder(override val containerView: View) : RecyclerView.ViewHolde
...
+        event.eventType.let {
+            if (it != null)
+                addchips(it.name)
+        }
+        event.eventTopic.let {
+            if (it != null)
+                addchips(it.name)
+        }
+        event.eventSubTopic.let {
+            if (it != null)
```
Comment (nikit19) [inline app/src/main/java/org/fossasia/openevent/general/event/EventViewHolder.kt]: any particular reason for explicit null checks?
Using `?.let` would work?
